src/crimson/net/SocketMessenger.cc

```diff
@@ -86,26 +81,31 @@ SocketMessenger::try_bind(const entity_addrvec_t& addrs,
     return bind(addrs);
   }
   ceph_assert(min_port <= max_port);
-  return seastar::do_with(uint32_t(min_port),
-    [this, max_port, addr] (auto& port) {
-      return seastar::repeat([this, max_port, addr, &port] {
-          auto to_bind = addr;
-          to_bind.set_port(port);
-          return bind(entity_addrvec_t{to_bind})
-            .then([this] {
-              logger().info("{}: try_bind: done", *this);
-              return stop_t::yes;
-            }).handle_exception_type([this, max_port, &port] (const std::system_error& e) {
-              ceph_assert(e.code() == error::address_in_use);
-              logger().trace("{}: try_bind: {} already used", *this, port);
-              if (port == max_port) {
-                throw e;
-              }
-              ++port;
-              return stop_t::no;
-            });
-        });
+  return seastar::do_with(uint32_t(min_port), false,
+      [this, max_port, addr] (auto& port, bool& bound) {
+    return seastar::repeat([this, max_port, addr, &port, &bound] {
+      auto to_bind = addr;
+      to_bind.set_port(port);
+      return bind(entity_addrvec_t{to_bind}
+      ).then([this, max_port, &port, &bound] (bool ret) {
+        if (ret) {
+          logger().info("{}: try_bind: done", *this);
+          bound = true;
+          return stop_t::yes;
+        } else {
+          logger().trace("{}: try_bind: {} cannot be used", *this, port);
+          if (port == max_port) {
+            return stop_t::yes;
```

Comment:
i think it'd be more like an exception -- something not expected.

Reply:
fixed, see https://github.com/ceph/ceph/pull/32632/commits/daa3cc2375369dcf0f9e5cbfdcded78e6a86b30e